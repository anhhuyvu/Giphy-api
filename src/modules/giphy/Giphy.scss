.gif-page-container {
  .spinner-container {
    display: flex;
    justify-content: center;
    top: 50%;
    position: absolute;
    left: 50%;
  }
  .giphy-container {
    display: grid;
    .img-container {
      img {
        margin-top: 20px;
        cursor: pointer;
        width: 50%;
        margin: 0 auto;
      }
    }

    @media (min-width: 992px) {
      grid-template-columns: auto auto auto auto;
      margin-left: 10%;
    }
    @media (min-width: 768px) and (max-width: 992px) {
      grid-template-columns: auto auto auto;
      margin-left: 15%;
    }
    @media (max-width: 768px) {
      grid-template-columns: auto auto;
      margin-left: 20%;
    }
  }
}
