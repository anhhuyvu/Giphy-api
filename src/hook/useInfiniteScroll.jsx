import { useState, useEffect } from 'react';
import { LIMIT_ITEMS, getGifItems } from '../api/giphy';

const useInfiniteScroll = (page) => {
  const [data, setData] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [hasMore, setHasMore] = useState(false);

  const fetchDataByPage = async (page) => {
    setIsLoading(true);
    const offset = (page - 1) * LIMIT_ITEMS;
    const response = await getGifItems(offset);

    const { data, pagination } = response;
    if (data && pagination) {
      setData((prevItem) => [...prevItem, ...data]);
      setIsLoading(false);
      setHasMore(page < pagination.total_count);
    }
  };

  useEffect(() => {
    fetchDataByPage(page);
  }, [page]);

  return { isLoading, data, hasMore };
};

export default useInfiniteScroll;
