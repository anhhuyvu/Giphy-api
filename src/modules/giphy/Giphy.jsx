import { useCallback, useRef, useState } from 'react';
import useInfiniteScroll from '../../hook/useInfiniteScroll';

import GifContainer from '../../components/gifContainer/GifContainer';
import Spinner from '../../components/spinner/Spinner';
import FullScreenImage from '../../components/fullScreenImage/FullScreenImage';

import './Giphy.scss';

const Giphy = () => {
  const [page, setPage] = useState(1);
  const lastElementObserver = useRef(null);
  const { data, isLoading, hasMore } = useInfiniteScroll(page);
  const [overlay, setOverlay] = useState(false);
  const [selectedGif, setSelectedGif] = useState('');

  const lastElementRef = useCallback(
    (element) => {
      if (isLoading) return;
      if (lastElementObserver.current) {
        lastElementObserver.current.disconnect();
      }
      lastElementObserver.current = new IntersectionObserver((items) => {
        if (items[0].isIntersecting && hasMore) {
          setPage((prevPage) => prevPage + 1);
        }
      });

      if (element) {
        lastElementObserver.current.observe(element);
      }
    },
    [isLoading, hasMore],
  );
  const toggle = () =>
    setOverlay((prevState) => {
      if (prevState) {
        setSelectedGif('');
      }
      return !prevState;
    });
  const handleOnClick = (url) => {
    setSelectedGif(url);
    toggle();
  };

  return (
    <div className="gif-page-container">
      {isLoading && (
        <div className="spinner-container">
          <Spinner />
        </div>
      )}
      <h1 className="d-flex justify-content-center">
        <b>Trending Gifs</b>
      </h1>
      <div className="giphy-container">
        {data.map((items, index) => {
          const { id, images } = items;
          const { url, title } = images.downsized;
          if (data.length - 1 === index) {
            return (
              <GifContainer key={id} url={url} title={title} ref={lastElementRef} onclick={() => handleOnClick(url)} />
            );
          }

          return <GifContainer key={id} url={url} title={title} onclick={() => handleOnClick(url)} />;
        })}
      </div>
      <FullScreenImage url={selectedGif} isOpen={overlay} toggle={toggle} />
    </div>
  );
};
export default Giphy;
