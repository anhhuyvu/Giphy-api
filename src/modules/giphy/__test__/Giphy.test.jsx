import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import Giphy from '../Giphy';

import * as useInfiniteScroll from '../../../hook/useInfiniteScroll';

jest.mock('../../../components/gifContainer/GifContainer', () => 'GifContainer');

const mockData = [
  {
    id: 'gif1',
    user: {
      display_name: 'name1',
      avatar_url: 'avatar1',
    },
    images: {
      downsized: {
        url: 'url1',
        title: 'title1',
      },
    },
  },
  {
    id: 'gif2',
    user: {
      display_name: 'name2',
      avatar_url: 'avatar2',
    },
    images: {
      downsized: {
        url: 'url2',
        title: 'title2',
      },
    },
  },
  {
    id: 'gif3',
    user: {
      display_name: 'name3',
      avatar_url: 'avatar3',
    },
    images: {
      downsized: {
        url: 'url3',
        title: 'title3',
      },
    },
  },
];

describe('test snapshot', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    window.IntersectionObserver = class {
      observe() {}
      disconnect() {}
    };
  });

  it('render loading state', () => {
    jest.spyOn(useInfiniteScroll, 'default').mockReturnValue({
      data: [],
      isLoading: true,
      lastPage: false,
      error: false,
    });

    const tree = renderer.create(<Giphy />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render cards', () => {
    jest.spyOn(useInfiniteScroll, 'default').mockReturnValue({
      data: mockData,
      isLoading: false,
      lastPage: false,
      error: false,
    });

    const tree = renderer.create(<Giphy />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('test react render', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    window.IntersectionObserver = class {
      observe() {}
      disconnect() {}
    };
  });

  it('test loading state', () => {
    jest.spyOn(useInfiniteScroll, 'default').mockReturnValue({
      data: mockData,
      isLoading: true,
      lastPage: false,
      error: false,
    });

    const wrapper = render(<Giphy />);
    expect(wrapper.container.querySelector('.spinner-container')).not.toBeNull();
  });

  it('test normal render', () => {
    jest.spyOn(useInfiniteScroll, 'default').mockReturnValue({
      data: mockData,
      isLoading: false,
      lastPage: false,
      error: false,
    });

    const wrapper = render(<Giphy />);
    expect(wrapper.container.querySelector('.spinner-container')).toBeNull();
  });

  it('test render when error occurred', () => {
    jest.spyOn(useInfiniteScroll, 'default').mockReturnValue({
      data: mockData,
      isLoading: false,
      lastPage: false,
      error: true,
    });

    const wrapper = render(<Giphy />);
    expect(wrapper.container.querySelector('.loading')).toBeNull();
  });
});
