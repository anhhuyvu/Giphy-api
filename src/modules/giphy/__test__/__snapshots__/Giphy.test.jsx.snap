// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot render cards 1`] = `
<div
  className="gif-page-container"
>
  <h1
    className="d-flex justify-content-center"
  >
    <b>
      Trending Gifs
    </b>
  </h1>
  <div
    className="giphy-container"
  >
    <GifContainer
      onclick={[Function]}
      title="title1"
      url="url1"
    />
    <GifContainer
      onclick={[Function]}
      title="title2"
      url="url2"
    />
    <GifContainer
      onclick={[Function]}
      title="title3"
      url="url3"
    />
  </div>
  <div
    className="fullscreen-container "
    id="fullscreen-container"
  >
    <img
      alt=""
      src=""
    />
  </div>
</div>
`;

exports[`test snapshot render loading state 1`] = `
<div
  className="gif-page-container"
>
  <div
    className="spinner-container"
  >
    <div
      className="lds-roller"
    >
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
  <h1
    className="d-flex justify-content-center"
  >
    <b>
      Trending Gifs
    </b>
  </h1>
  <div
    className="giphy-container"
  />
  <div
    className="fullscreen-container "
    id="fullscreen-container"
  >
    <img
      alt=""
      src=""
    />
  </div>
</div>
`;
